pkg/daemon/daemon.go

```diff
@@ -498,60 +505,51 @@ func (dn *Daemon) triggerUpdate() error {
 	return dn.triggerUpdateWithMachineConfig(nil)
 }
 
-// isDesiredMachineState confirms that the node is actually in the state that it
-// wants to be in. It does this by looking at the elements in the target config
-// and checks if all are present on the node. Returns true iff there are no
-// mismatches (e.g. files, units, OS version), as well as the config that was
-// evaluated if the state is what the machine wants to be in.
-func (dn *Daemon) isDesiredMachineState() (bool, string, error) {
+// validateOnDiskState compares the on-disk state against what a configuration
+// specifies.  If for example an admin ssh'd into a node, or another operator
+// is stomping on our files, we want to highlight that and mark the system
+// degraded.
+func (dn *Daemon) validateOnDiskState(currentConfig *mcfgv1.MachineConfig) bool {
+	if !dn.checkFiles(currentConfig.Spec.Config.Storage.Files) {
+		return false
+	}
+	if !dn.checkUnits(currentConfig.Spec.Config.Systemd.Units) {
+		return false
+	}
+	return true
+}
+
+// isDesiredMachineState looks at the node annotations and determines if
+// our current state matches the desired one.
+func (dn *Daemon) isDesiredMachineState() (bool, *mcfgv1.MachineConfig, error) {
 	ccAnnotation, err := getNodeAnnotation(dn.kubeClient.CoreV1().Nodes(), dn.name, CurrentMachineConfigAnnotationKey)
 	if err != nil {
-		return false, "", err
+		return false, nil, err
 	}
 	dcAnnotation, err := getNodeAnnotation(dn.kubeClient.CoreV1().Nodes(), dn.name, DesiredMachineConfigAnnotationKey)
 	if err != nil {
-		return false, "", err
+		return false, nil, err
 	}
 
 	currentConfig, err := getMachineConfig(dn.client.MachineconfigurationV1().MachineConfigs(), ccAnnotation)
 	if err != nil {
-		return false, "", err
+		return false, nil, err
 	}
+	currentConfigName := currentConfig.GetName()
 	desiredConfig, err := getMachineConfig(dn.client.MachineconfigurationV1().MachineConfigs(), dcAnnotation)
 	if err != nil {
-		return false, "", err
-	}
-
-	// if we can't reconcile the changes between the old config and the new
-	// config, the machine is definitely not in its desired state. this function
-	// will return true (meaning they are reconcilable) if there aren't actually
-	// changes.
-	reconcilableError := dn.reconcilable(currentConfig, desiredConfig)
-	if reconcilableError != nil {
-		return false, "", nil
+		return false, nil, err
 	}
+	desiredConfigName := desiredConfig.GetName()
 
-	isDesiredOS := false
-	// We only deal with operating system management on RHCOS
-	if dn.OperatingSystem != MachineConfigDaemonOSRHCOS {
-		// If we are on anything but RHCOS we set to True as there
-		// is nothing to updated.
-		isDesiredOS = true
+	if currentConfigName != desiredConfigName {
```

Comment:
I'm not sure I follow this. Won't this simpler logic result in a reboot loop?

- a new desired config node annotation is set
- the daemon sees it and applies the update and reboots the system
- the daemon comes back online after the reboot and goes through here
- `currentConfigName != desiredConfigName` evaluates to true since the node annotation hasn't been updated
- daemon tries to apply the same update again

We need to only be caring about the desired config, not the current config, because this is how we determine whether we should update the current config annotation itself after an update.

Reply:
Yep, I have a followup in https://github.com/openshift/machine-config-operator/pull/245#issuecomment-448017388

Reply:
I pushed my WIP on top of this below :arrow_down: so you all can see what I'm thinking.